test/extensions/filters/http/json_to_metadata/integration_test.cc

```diff
@@ -0,0 +1,198 @@
+  Http::TestRequestTrailerMapImpl incoming_trailers_{{"request1", "trailer1"},
+                                                     {"request2", "trailer2"}};
+
+  const std::string request_body_ =
+      R"delimiter(
+        {"version":"1.0.0",
+        "messages":[
+          {"role":"user","content":"content A"},
+          {"role":"assistant","content":"content B"},
+          {"role":"user","content":"content C"},
+          {"role":"assistant","content":"content D"},
+          {"role":"user","content":"content E"}],
+        "stream":true})delimiter";
+};
+
+// TODO(#26236): Fix test suite for HTTP/3.
+INSTANTIATE_TEST_SUITE_P(
+    Protocols, JsonToMetadataIntegrationTest,
+    testing::ValuesIn(HttpProtocolIntegrationTest::getProtocolTestParamsWithoutHTTP3()),
+    HttpProtocolIntegrationTest::protocolTestParamsToString);
+
+TEST_P(JsonToMetadataIntegrationTest, Basic) {
+  initializeFilter();
+
+  runTest(incoming_headers_, request_body_);
+
+  EXPECT_EQ(1UL, test_server_->counter("json_to_metadata.rq_success")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_mismatched_content_type")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_no_body")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_invalid_json_body")->value());
+}
+
+TEST_P(JsonToMetadataIntegrationTest, BasicOneChunk) {
+  initializeFilter();
+
+  runTest(incoming_headers_, request_body_, 1);
+
+  EXPECT_EQ(1UL, test_server_->counter("json_to_metadata.rq_success")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_mismatched_content_type")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_no_body")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_invalid_json_body")->value());
+}
+
+TEST_P(JsonToMetadataIntegrationTest, Trailer) {
+  initializeFilter();
+
+  runTest(incoming_headers_, request_body_, 5, true);
+
+  EXPECT_EQ(1UL, test_server_->counter("json_to_metadata.rq_success")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_mismatched_content_type")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_no_body")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_invalid_json_body")->value());
+}
+
+TEST_P(JsonToMetadataIntegrationTest, GoOverWaterMarkAtEndStream) {
+  config_helper_.setBufferLimits(request_body_.length(), request_body_.length());
+  initializeFilter();
+
+  runTest(incoming_headers_, request_body_);
+
+  EXPECT_EQ(1UL, test_server_->counter("json_to_metadata.rq_success")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_mismatched_content_type")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_no_body")->value());
+  EXPECT_EQ(0UL, test_server_->counter("json_to_metadata.rq_invalid_json_body")->value());
+}
+
+TEST_P(JsonToMetadataIntegrationTest, GoOverWaterMarkBeforeEndStream) {
```

Comment:
Test is failing in arm, and windows, looks legit:

```
[ RUN      ] Protocols/JsonToMetadataIntegrationTest.GoOverWaterMarkBeforeEndStream/IPv4_HttpDownstream_HttpUpstreamHttpParserNghttp2NoDeferredProcessing
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0xbeb39f8000000010
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.28.0-dev/test/RELEASE/BoringSSL
[2023-08-04 05:43:16.081][62580][critical][backtrace] [./source/server/backtrace.h:96] #0: __kernel_rt_sigreturn [0xffffb8aad78c]
[2023-08-04 05:43:16.092][62580][critical][backtrace] [./source/server/backtrace.h:96] #1: Envoy::IntegrationCodecClient::sendData() [0xa8eb80]
[2023-08-04 05:43:16.103][62580][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::(anonymous namespace)::JsonToMetadataIntegrationTest::runTest() [0xa6f160]
[2023-08-04 05:43:16.114][62580][critical][backtrace] [./source/server/backtrace.h:96] #3: Envoy::(anonymous namespace)::JsonToMetadataIntegrationTest_GoOverWaterMarkBeforeEndStream_Test::TestBody() [0xa72538]
[2023-08-04 05:43:16.125][62580][critical][backtrace] [./source/server/backtrace.h:96] #4: testing::internal::HandleExceptionsInMethodIfSupported<>() [0x1caa7b8]
```

Reply:
It crashes in the testing framework here with dangling raw point `inner_encoder_`
https://github.com/envoyproxy/envoy/blob/b1ea06265c05799f4750f577090d2a33d5829fd8/source/common/http/codec_wrappers.h#L82

```
(gdb) p inner_encoder_
$4 = (Envoy::Http::RequestEncoder *) 0x2578bd512570
(gdb) info vtbl inner_encoder_
warning: can't find linker symbol for virtual table for `Envoy::Http::RequestEncoder' value
```

I suspect I abused the test framework :P 

Moreover, I substitue the prepended filter by `wait-for-whole-request-and-response-filter`
to simulate the buffer behavior and it still crashes, which kinda proves the crash is not related to
json to metadata filter. 

I'm going to remove the unnecessary test to move forward.